libsignal-protocol/src/store/mod.rs

```diff
@@ -0,0 +1,84 @@
+pub(crate) mod identity_key_store;
+pub(crate) mod pre_key_store;
+pub(crate) mod session_store;
+pub(crate) mod signed_pre_key_store;
+
+pub use self::{
+    identity_key_store::IdentityKeyStore,
+    pre_key_store::PreKeyStore,
+    session_store::{SerializedSession, SessionStore},
+    signed_pre_key_store::SignedPreKeyStore,
+};
+
+#[cfg(feature = "in-memory-store")]
+pub mod in_memory_store {
+    use crate::{keys::IdentityKeyPair, Address, InternalError};
+    use std::{borrow::ToOwned, collections::HashMap, sync::Mutex};
+
+    pub struct InMemoryIdentityKeyStore<'a> {
+        trusted_keys: Mutex<HashMap<Address<'a>, Vec<u8>>>,
+        local_registration_id: u32,
+        identity_key_pair: IdentityKeyPair,
+    }
+
+    impl<'a> InMemoryIdentityKeyStore<'a> {
+        pub fn new(
+            identity_key_pair: IdentityKeyPair,
+            local_registration_id: u32,
+        ) -> Self {
+            Self {
+                trusted_keys: Mutex::new(HashMap::new()),
+                local_registration_id,
+                identity_key_pair,
+            }
+        }
+    }
+
+    impl<'a> super::IdentityKeyStore<'a> for InMemoryIdentityKeyStore<'a> {
+        fn identity_key_pair(&self) -> Result<IdentityKeyPair, InternalError> {
+            Ok(self.identity_key_pair.clone())
+        }
+
+        fn local_registration_id(&self) -> Result<u32, InternalError> {
+            Ok(self.local_registration_id)
+        }
+
+        fn save_identity(
+            &self,
+            address: Address<'a>,
+            identity_key: &[u8],
+        ) -> Result<bool, InternalError> {
+            let mut guard = self
+                .trusted_keys
+                .lock()
+                .map_err(|_| InternalError::Unknown)?;
+            let existing = guard.contains_key(&address);
+            if !existing {
+                guard.insert(address, identity_key.to_owned());
+                Ok(true)
+            } else {
+                Ok(false)
+            }
+        }
+
+        fn is_trusted_identity(
+            &self,
+            address: Address,
+            identity_key: &[u8],
+        ) -> Result<bool, InternalError> {
+            let guard = self
+                .trusted_keys
+                .lock()
+                .map_err(|_| InternalError::Unknown)?;
+            let identity = guard.get(&address);
+            let is_trusted = identity.is_none() || {
```

Comment:
This looks a little funny. Could we use an early return here (e.g. `if identity.is_none() { return Ok(true); }`) to reduce nesting? Otherwise a simple match would be okay.

```rust
match identity.is_none() {
  Some(data) if data == identity_key => Ok(true),
  Some(_) | None => Ok(false),
}
```

Reply:
Note that the dummy `IdentityKeyStore` in our helpers module will trust an identity we don't know (the `identity.is_none()`). It was in the original C library's tests, so I'm assuming that's just to make testing easier (you can skip the whole "trust" process). 

We probably don't want that in our in-memory implementation because otherwise any random person we've never talked to would be marked as "trusted" the first time they contact us.

Reply:
I used the documentation from the original java library:
https://github.com/signalapp/libsignal-protocol-java/blob/9518b127cee2b985293e79f149ad64309bc17688/java/src/main/java/org/whispersystems/libsignal/state/IdentityKeyStore.java#L52-L71

So i thought it is the right thing to do .

Reply:
Ah, good point. If that's the case then we'll go with that behaviour.

Would you be able to mention that in the in-memory store type's doc-comment?

> Convention is that the Signal Protocol is 'trust on first use.'  This means that an identity key is onsidered 'trusted' if there is no entry for the recipient in the local store, or if it matches the saved key for a recipient in the local store.  Only if it mismatches an entry in the local store is it considered 'untrusted.'